New resource aws_glue_dev_endpoint

jckuester opened the pull request:
Changes proposed in this pull request:

* Add support for new resource `aws_glue_dev_endpoint`

Output from acceptance testing:

```$ make testacc TESTARGS='-run=TestAccGlueDevEndpoint_*'
--- PASS: TestAccGlueDevEndpoint_ExtraPythonLibsS3Path (506.02s)
--- PASS: TestAccGlueDevEndpoint_Basic (532.90s)
--- PASS: TestAccGlueDevEndpoint_SecurityConfiguration (546.74s)
--- PASS: TestAccGlueDevEndpoint_NumberOfWorkers (556.80s)
--- PASS: TestAccGlueDevEndpoint_Arguments (590.41s)
--- PASS: TestAccGlueDevEndpoint_Tags (598.73s)
--- PASS: TestAccGlueDevEndpoint_PublicKey (608.57s)
--- PASS: TestAccGlueDevEndpoint_PublicKeys (797.27s)
--- PASS: TestAccGlueDevEndpoint_ExtraJarsS3Path (853.46s)
--- PASS: TestAccGlueDevEndpoint_GlueVersion (948.85s)
--- PASS: TestAccGlueDevEndpoint_NumberOfNodes (972.99s)
--- PASS: TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs (1194.39s)
--- PASS: TestAccGlueDevEndpoint_WorkerType (1444.21s)
--- PASS: TestAccGlueDevEndpoint_disappears (472.50s)
...
```

xmj commented:
@jckuester wondering if you could rebase this and fix the merge conflict

maartenpeters commented:
Is there any indication as to when this feature will be merged?

kabeersvohra commented:
Can this be reviewed? Thanks

aeschright commented:
Hi folks! The AWS Provider team is still working on catching up to a substantial PR backlog. You can see more about the process here: https://github.com/terraform-providers/terraform-provider-aws/wiki/Frequently-Asked-Questions#when-will-this-pull-request-be-merged

We'll have an update on the PR as soon as we can. Thanks for the work you've done so far.

jckuester commented:
I updated this PR to support then newest API changes from AWS, ie. I added support for the  glue_version, worker_type, etc. arguments. Tests are all passing.

DrFaust92 commented:
Closes #3876

clockworksoul commented:
What's the status of this request?

teamterraform commented:
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Action Required: Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

jckuester commented:
Thanks for the info. I'll adapt my code in this PR accordingly to the new requirements.

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,945 @@
+
+					testAccCheckResourceAttrRegionalARN(resourceName, "arn", "glue", fmt.Sprintf("devEndpoint/%s", rName)),
+					resource.TestCheckResourceAttr(resourceName, "name", rName),
+					testAccMatchResourceAttrGlobalARN(resourceName, "role_arn", "iam", regexp.MustCompile(`role/AWSGlueServiceRole-tf-acc-test-[0-9]+$`)),
```
lets swap this for `					resource.TestCheckResourceAttrPair(resourceName, "role_arn", roleResourceName, "role_arn"),
`

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,945 @@
+				Config: testAccGlueDevEndpointConfig_Arguments(rName, "baz"),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
remove empty line

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,945 @@
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
```
lets add another test case where it adds another argument.

jckuester replied:
Good idea. I also cleaned up the test to have a similar structure as `TestAccAWSGlueJob_Tags`.

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,945 @@
+				Config: testAccGlueDevEndpointConfig_ExtraJarsS3Path(rName, extraJarsS3Path),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
empty line

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,945 @@
+				Config: testAccGlueDevEndpointConfig_ExtraJarsS3Path(rName, extraJarsS3PathUpdated),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
empty line

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,945 @@
+				Config: testAccGlueDevEndpointConfig_ExtraPythonLibsS3Path(rName, extraPythonLibsS3Path),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
empty line

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,945 @@
+				Config: testAccGlueDevEndpointConfig_ExtraPythonLibsS3Path(rName, extraPythonLibsS3PathUpdated),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
empty line

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,616 @@
+				ForceNew: true,
+			},
+
+			"security_group_ids": {
```
lets add `RequiredWith: []string{"subnet_id"}` here

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,616 @@
+				Set:      schema.HashString,
+			},
+
+			"subnet_id": {
```
same as security groups

DrFaust92 reviewed (changes requested):
Thanks for this!
some comments below.

can you also get rid of the empty lines in the test configs? thanks!

jckuester commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,616 @@
+		oldRaw, newRaw := d.GetChange("public_keys")
+		old := oldRaw.(*schema.Set)
+		new := newRaw.(*schema.Set)
+		create, remove := diffPublicKeys(expandStringSet(old), expandStringSet(new))
```
I changed it and removed `diffPublicKeys`. What's about the other custom function `diffArguments` that I added?

DrFaust92 commented on aws/publicKeys_test.go:
```diff
@@ -0,0 +1,67 @@
+package aws
```
this files needs to be removed

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,580 @@
+	var create = make(map[string]*string)
+	var remove []*string
+
+	for oldArgKey, oldArgVal := range oldArgs {
```
instead of for in for:

pseudocode:

```
for oldK, oldV in old_map {
  if val, ok := new_map[oldK]; ok {
    if val != oldV {
      removed = append(removed, old)
    }
  } else {
   removed = append(removed, old)
}

for newK, newV in new_map {
  if _, ok := dict[newK]; !ok {
    create[newK] = newV
  }
}
```

DrFaust92 reviewed (changes requested):
some changes, im not aware of an existing function to diff maps but see my implementation below.

DrFaust92 commented:
@jckuester, can you also add a disappears test per https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/contributing/running-and-writing-acceptance-tests.md#disappears-acceptance-tests

jckuester commented:
@DrFaust92 thanks. I'll let you know, when I am done addressing your feedback. I've a busy week, so probably will need a few more days.

jckuester commented:
@DrFaust92 I am done addressing the feedback. Ready for a 2nd review round!

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,588 @@
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
```
switch to:

```
validation.StringMatch(regexp.MustCompile(`^\w+\.\w+$`), "must match version pattern X.X"),
```

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,588 @@
+func resourceAwsGlueDevEndpointCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).glueconn
+
+	var name string
+	if v, ok := d.GetOk("name"); ok {
+		name = v.(string)
+	} else {
+		name = resource.UniqueId()
+	}
```
not sure about this, id rather match the API and set this to required and lose the unique name generator

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,588 @@
+	d.SetId(name)
+
+	log.Printf("[DEBUG] Waiting for Glue Dev Endpoint (%s) to become available", d.Id())
+	stateConf := &resource.StateChangeConf{
```
now that the refresh function is extracted, lets extract this waiter as well to the glue waiter package(see other glue functions in that package)

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,588 @@
+
+	log.Printf("[INFO] Deleting Glue Dev Endpoint: %s", d.Id())
+
+	_, err := conn.DeleteDevEndpoint(deleteOpts)
```
may worth to consider a waiter for delete, according to docs there is a Terminating status: https://docs.aws.amazon.com/glue/latest/dg/console-development-endpoint.html

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_GlueVersion(rName, "1.0"),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

jckuester replied:
Sorry, I hope I caught now all the spaces. Would probably be good to have a linter for this at some point similar to: https://github.com/bombsimon/wsl

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_GlueVersion(rName, "0.9"),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_NumberOfNodes(rName, 2),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_NumberOfNodes(rName, 5),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_NumberOfWorkers(rName, 2),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_NumberOfWorkers(rName, 5),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_PublicKey(rName, publicKey1),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_PublicKey(rName, publicKey2),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_PublicKeys2(rName, publicKey1, publicKey2),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_PublicKeys3(rName, publicKey1, publicKey3, publicKey4),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_SecurityConfiguration(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_SubnetID_SecurityGroupIDs(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_WorkerType_Standard(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_WorkerType(rName, glue.WorkerTypeG1x),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+				Config: testAccGlueDevEndpointConfig_WorkerType(rName, glue.WorkerTypeG2x),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSGlueDevEndpointExists(resourceName, &endpoint),
+
```
extraneous space

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+func testAccGlueDevEndpointConfig_Base(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_iam_role" "test" {
+  name               = "AWSGlueServiceRole-%s"
```
lets use `%[1]s` and `%[1]q` for the one below and use `rName` once in the fmt.Sprintf params

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+    }
+  }
+}
+`, rName, rName)
```
same as above use `%[1]q` for both occurrences

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+  }
+  depends_on = [aws_iam_role_policy_attachment.glue_service_role]
+}
+`, rName, rName, rName, rName)
```
same as above

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+  worker_type       = %[2]q
+  number_of_workers = 2
+}
+`, rName, glue.WorkerTypeStandard)
```
lets hardcode this in the config if its not being passed

DrFaust92 commented on website/docs/r/glue_dev_endpoint.markdown:
````diff
@@ -0,0 +1,86 @@
+Basic usage:
+
+```hcl
+resource "aws_glue_dev_endpoint" "de" {
````
lets use `example` for the resource names and references

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,968 @@
+}
+
+resource "aws_iam_role_policy_attachment" "glue_service_role" {
+  policy_arn = "arn:aws:iam::aws:policy/service-role/AWSGlueServiceRole"
```
lets add

`data "aws_partition" "current" {}`

and use `arn:${data.aws_partition.current.partition}:iam::aws:policy/service-role/AWSGlueServiceRole` so test could run in other partitions as well.

DrFaust92 reviewed (changes requested):
Round 2, mostly cosmetic changes.

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,588 @@
+	customLibs := &glue.DevEndpointCustomLibraries{}
+
+	if d.HasChange("arguments") {
+		oldRaw, newRaw := d.GetChange("arguments")
+		old := oldRaw.(map[string]interface{})
+		new := newRaw.(map[string]interface{})
+		create, remove := diffArguments(old, new)
+		input.AddArguments = create
+		input.DeleteArguments = remove
+
+		hasChanged = true
```
found that `diffStringMaps` function exists for this!

jckuester commented:
@DrFaust92 thanks for 2nd feedback round. I will find time in the next few days to address the new feedback and will let you know when done.

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,529 @@
+	}
+
+	if kv, ok := d.GetOk("arguments"); ok {
+		arguments := make(map[string]string)
+		for k, v := range kv.(map[string]interface{}) {
+			arguments[k] = v.(string)
+		}
+		input.Arguments = aws.StringMap(arguments)
```
input.Arguments = stringMapToPointers(ec.(map[string]interface{}))

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,529 @@
+	}
+
+	log.Printf("[DEBUG] Creating Glue Dev Endpoint: %#v", *input)
+	err := resource.Retry(1*time.Minute, func() *resource.RetryError {
```
err := resource.Retry(iamwaiter.PropagationTimeout, func() *resource.RetryError {

DrFaust92 replied:
import 
`iamwaiter "github.com/terraform-providers/terraform-provider-aws/aws/internal/service/iam/waiter"`

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,529 @@
+		oldRaw, newRaw := d.GetChange("arguments")
+		old := oldRaw.(map[string]interface{})
+		new := newRaw.(map[string]interface{})
+		create, remove := diffStringMaps(old, new)
```
create, remove, _ := diffStringMaps(old, new)

DrFaust92 replied:
see CI fail

jckuester replied:
I see, the 3 return values were a recent changes, I haven't rebased yet.

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint.go:
```diff
@@ -0,0 +1,529 @@
+		}
+		os := o.(*schema.Set)
+		ns := n.(*schema.Set)
+		remove := os.Difference(ns).List()
```
no need to call `List()`

use `expandStringSet(remove)`

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,954 @@
+}
+
+func testAccGlueDevEndpointConfig_SubnetID_SecurityGroupIDs(rName string) string {
+	return testAccGlueDevEndpointConfig_Base(rName) + fmt.Sprintf(`
```
return composeConfig(testAccAvailableAZsNoOptInConfig(), testAccGlueDevEndpointConfig_Base(rName), fmt.Sprintf(`

DrFaust92 commented on aws/resource_aws_glue_dev_endpoint_test.go:
```diff
@@ -0,0 +1,954 @@
+resource "aws_subnet" "test" {
+  vpc_id            = aws_vpc.test.id
+  cidr_block        = "10.0.1.0/24"
+  availability_zone = "us-west-2a"
```
`availability_zone = data.aws_availability_zones.available.names[0]`

DrFaust92 reviewed (changes requested):
Some more comments, I believe this is it(except CI fail on map diff functions)

jckuester commented:
Almost done, I just need to debug test errors when waiting for the endpoints to terminate:

```
2020/11/03 16:08:55 [WARN] Got error running Terraform: 2020/11/03 16:06:01 [DEBUG] Using modified User-Agent: Terraform/0.12.28 HashiCorp-terraform-exec/0.10.0

Error: error while waiting for Glue Dev Endpoint (tf-acc-test-166304744643581240) to become terminated: couldn't find resource (21 retries)


    testing_new.go:62: Error running post-test destroy, there may be dangling resources: 2020/11/03 16:06:01 [DEBUG] Using modified User-Agent: Terraform/0.12.28 HashiCorp-terraform-exec/0.10.0
        
        Error: error while waiting for Glue Dev Endpoint (tf-acc-test-166304744643581240) to become terminated: couldn't find resource (21 retries)
```

jckuester commented:
@DrFaust92 I am done addressing feedback from the 2nd round.

DrFaust92 reviewed (approved):
LGTM

```
--- PASS: TestAccGlueDevEndpoint_disappears (503.34s)--- PASS: TestAccGlueDevEndpoint_ExtraPythonLibsS3Path (517.23s)
--- PASS: TestAccGlueDevEndpoint_Tags (591.90s)
--- PASS: TestAccGlueDevEndpoint_ExtraJarsS3Path (596.16s)
--- PASS: TestAccGlueDevEndpoint_Arguments (603.62s)
--- PASS: TestAccGlueDevEndpoint_SecurityConfiguration (649.98s)
--- PASS: TestAccGlueDevEndpoint_Basic (660.52s)
--- PASS: TestAccGlueDevEndpoint_PublicKeys (666.27s)
--- PASS: TestAccGlueDevEndpoint_PublicKey (710.84s)
--- PASS: TestAccGlueDevEndpoint_NumberOfNodes (1028.32s)
--- PASS: TestAccGlueDevEndpoint_GlueVersion (1033.62s)
--- PASS: TestAccGlueDevEndpoint_NumberOfWorkers (1109.88s)
--- PASS: TestAccGlueDevEndpoint_WorkerType (1119.71s)
--- PASS: TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs (1201.81s)
```

jckuester commented:
I had to rebase one more time, because the changes I made to fix the linter issue in `/aws/resource_aws_kinesis_firehose_delivery_stream_test.go` has been merged to master and caused a merge conflict.

breathingdust commented:
LGTM :rocket: Thanks @jckuester for the implementation

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccGlueDevEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccGlueDevEndpoint_ -timeout 120m
=== RUN   TestAccGlueDevEndpoint_Basic
=== PAUSE TestAccGlueDevEndpoint_Basic
=== RUN   TestAccGlueDevEndpoint_Arguments
=== PAUSE TestAccGlueDevEndpoint_Arguments
=== RUN   TestAccGlueDevEndpoint_ExtraJarsS3Path
=== PAUSE TestAccGlueDevEndpoint_ExtraJarsS3Path
=== RUN   TestAccGlueDevEndpoint_ExtraPythonLibsS3Path
=== PAUSE TestAccGlueDevEndpoint_ExtraPythonLibsS3Path
=== RUN   TestAccGlueDevEndpoint_GlueVersion
=== PAUSE TestAccGlueDevEndpoint_GlueVersion
=== RUN   TestAccGlueDevEndpoint_NumberOfNodes
=== PAUSE TestAccGlueDevEndpoint_NumberOfNodes
=== RUN   TestAccGlueDevEndpoint_NumberOfWorkers
=== PAUSE TestAccGlueDevEndpoint_NumberOfWorkers
=== RUN   TestAccGlueDevEndpoint_PublicKey
=== PAUSE TestAccGlueDevEndpoint_PublicKey
=== RUN   TestAccGlueDevEndpoint_PublicKeys
=== PAUSE TestAccGlueDevEndpoint_PublicKeys
=== RUN   TestAccGlueDevEndpoint_SecurityConfiguration
=== PAUSE TestAccGlueDevEndpoint_SecurityConfiguration
=== RUN   TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs
=== PAUSE TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs
=== RUN   TestAccGlueDevEndpoint_Tags
=== PAUSE TestAccGlueDevEndpoint_Tags
=== RUN   TestAccGlueDevEndpoint_WorkerType
=== PAUSE TestAccGlueDevEndpoint_WorkerType
=== RUN   TestAccGlueDevEndpoint_disappears
=== PAUSE TestAccGlueDevEndpoint_disappears
=== CONT  TestAccGlueDevEndpoint_Basic
=== CONT  TestAccGlueDevEndpoint_PublicKeys
=== CONT  TestAccGlueDevEndpoint_WorkerType
=== CONT  TestAccGlueDevEndpoint_ExtraPythonLibsS3Path
=== CONT  TestAccGlueDevEndpoint_Tags
=== CONT  TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs
=== CONT  TestAccGlueDevEndpoint_SecurityConfiguration
=== CONT  TestAccGlueDevEndpoint_GlueVersion
=== CONT  TestAccGlueDevEndpoint_PublicKey
=== CONT  TestAccGlueDevEndpoint_NumberOfWorkers
=== CONT  TestAccGlueDevEndpoint_NumberOfNodes
=== CONT  TestAccGlueDevEndpoint_disappears
=== CONT  TestAccGlueDevEndpoint_ExtraJarsS3Path
=== CONT  TestAccGlueDevEndpoint_Arguments
--- PASS: TestAccGlueDevEndpoint_SecurityConfiguration (512.71s)
--- PASS: TestAccGlueDevEndpoint_ExtraJarsS3Path (520.95s)
--- PASS: TestAccGlueDevEndpoint_disappears (528.26s)
--- PASS: TestAccGlueDevEndpoint_PublicKey (541.51s)
--- PASS: TestAccGlueDevEndpoint_ExtraPythonLibsS3Path (572.25s)
--- PASS: TestAccGlueDevEndpoint_PublicKeys (588.59s)
--- PASS: TestAccGlueDevEndpoint_Basic (648.55s)
--- PASS: TestAccGlueDevEndpoint_Tags (680.84s)
--- PASS: TestAccGlueDevEndpoint_Arguments (694.13s)
--- PASS: TestAccGlueDevEndpoint_GlueVersion (908.04s)
--- PASS: TestAccGlueDevEndpoint_NumberOfNodes (937.37s)
--- PASS: TestAccGlueDevEndpoint_NumberOfWorkers (958.03s)
--- PASS: TestAccGlueDevEndpoint_WorkerType (1167.59s)
--- PASS: TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs (1192.83s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1198.043s
```

GovCloud however has a consistent test failure (us-gov-west-1), #16170 tracks this issue for internal resolution.

```console
make testacc TEST=./aws/ TESTARGS='-run=TestAccGlueDevEndpoint_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccGlueDevEndpoint_ -timeout 120m
=== RUN   TestAccGlueDevEndpoint_Basic
=== PAUSE TestAccGlueDevEndpoint_Basic
=== RUN   TestAccGlueDevEndpoint_Arguments
=== PAUSE TestAccGlueDevEndpoint_Arguments
=== RUN   TestAccGlueDevEndpoint_ExtraJarsS3Path
=== PAUSE TestAccGlueDevEndpoint_ExtraJarsS3Path
=== RUN   TestAccGlueDevEndpoint_ExtraPythonLibsS3Path
=== PAUSE TestAccGlueDevEndpoint_ExtraPythonLibsS3Path
=== RUN   TestAccGlueDevEndpoint_GlueVersion
=== PAUSE TestAccGlueDevEndpoint_GlueVersion
=== RUN   TestAccGlueDevEndpoint_NumberOfNodes
=== PAUSE TestAccGlueDevEndpoint_NumberOfNodes
=== RUN   TestAccGlueDevEndpoint_NumberOfWorkers
=== PAUSE TestAccGlueDevEndpoint_NumberOfWorkers
=== RUN   TestAccGlueDevEndpoint_PublicKey
=== PAUSE TestAccGlueDevEndpoint_PublicKey
=== RUN   TestAccGlueDevEndpoint_PublicKeys
=== PAUSE TestAccGlueDevEndpoint_PublicKeys
=== RUN   TestAccGlueDevEndpoint_SecurityConfiguration
=== PAUSE TestAccGlueDevEndpoint_SecurityConfiguration
=== RUN   TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs
=== PAUSE TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs
=== RUN   TestAccGlueDevEndpoint_Tags
=== PAUSE TestAccGlueDevEndpoint_Tags
=== RUN   TestAccGlueDevEndpoint_WorkerType
=== PAUSE TestAccGlueDevEndpoint_WorkerType
=== RUN   TestAccGlueDevEndpoint_disappears
=== PAUSE TestAccGlueDevEndpoint_disappears
=== CONT  TestAccGlueDevEndpoint_Basic
=== CONT  TestAccGlueDevEndpoint_PublicKeys
=== CONT  TestAccGlueDevEndpoint_WorkerType
=== CONT  TestAccGlueDevEndpoint_SecurityConfiguration
=== CONT  TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs
=== CONT  TestAccGlueDevEndpoint_GlueVersion
=== CONT  TestAccGlueDevEndpoint_PublicKey
=== CONT  TestAccGlueDevEndpoint_NumberOfWorkers
=== CONT  TestAccGlueDevEndpoint_Tags
=== CONT  TestAccGlueDevEndpoint_NumberOfNodes
=== CONT  TestAccGlueDevEndpoint_ExtraJarsS3Path
=== CONT  TestAccGlueDevEndpoint_ExtraPythonLibsS3Path
=== CONT  TestAccGlueDevEndpoint_disappears
=== CONT  TestAccGlueDevEndpoint_Arguments
2020/11/12 13:48:58 [WARN] Truncating attribute path of 0 diagnostics for TypeSet
2020/11/12 13:49:00 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
2020/11/12 13:49:00 [DEBUG] Trying to get account information via sts:GetCallerIdentity
=== CONT  TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs
    resource_aws_glue_dev_endpoint_test.go:447: Step 1/2 error: Error running apply: 2020/11/12 13:49:01 [DEBUG] Using modified User-Agent: Terraform/0.12.29 HashiCorp-terraform-exec/0.10.0

        Error: error while waiting for Glue Dev Endpoint (tf-acc-test-831422846705845092) to become available: unexpected state 'FAILED', wanted target 'READY'. last error: %!s(<nil>)


--- PASS: TestAccGlueDevEndpoint_PublicKeys (429.68s)
--- PASS: TestAccGlueDevEndpoint_disappears (430.88s)
2020/11/12 13:56:10 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
--- PASS: TestAccGlueDevEndpoint_SecurityConfiguration (469.67s)
--- PASS: TestAccGlueDevEndpoint_Basic (477.31s)
--- PASS: TestAccGlueDevEndpoint_ExtraPythonLibsS3Path (481.32s)
--- PASS: TestAccGlueDevEndpoint_PublicKey (513.28s)
--- PASS: TestAccGlueDevEndpoint_Tags (546.82s)
--- PASS: TestAccGlueDevEndpoint_ExtraJarsS3Path (549.80s)
--- PASS: TestAccGlueDevEndpoint_Arguments (620.08s)
--- FAIL: TestAccGlueDevEndpoint_SubnetID_SecurityGroupIDs (690.11s)
--- PASS: TestAccGlueDevEndpoint_NumberOfNodes (776.66s)
--- PASS: TestAccGlueDevEndpoint_WorkerType (843.22s)
--- PASS: TestAccGlueDevEndpoint_GlueVersion (909.22s)
--- PASS: TestAccGlueDevEndpoint_NumberOfWorkers (1138.67s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	1140.605s
FAIL
make: *** [testacc] Error 1
```
